src/Model/Table/QueueProcessesTable.php

```diff
@@ -198,10 +198,13 @@ public function remove(string $pid): void {
 	 * @return int
 	 */
 	public function cleanEndedProcesses(): int {
-		$timeout = Config::defaultworkertimeout();
-		$thresholdTime = (new DateTime())->subSeconds($timeout);
+		$activeProcesses = $this->findActive()->all()->extract('id')->toArray();
+		$conditions = [];
+		if (count($activeProcesses) > 0) {
+			$conditions = ['id NOT IN' => $activeProcesses];
+		}
 
-		return $this->deleteAll(['modified <' => $thresholdTime]);
+		return $this->deleteAll($conditions);
```

Comment:
Instead of a 2nd query, cannot we just clean out the ones that are definitly done? 
So all the ones that are completed NOT NULL?

Reply:
what if we use the max runtime? as it couldnt be running any longer than that, correct?

Reply:
> Sry, I thought this was QueuedJobs table
> 
> But indeed, we probably still can keep a single query here, no?

Yes, but the unit test did not like the resulting SQL: see [commit](https://github.com/dereuromark/cakephp-queue/pull/401/commits/5e2ec5497b17e4e44ef44c4dcbec9d79bb955a92)

 
> what if we use the max runtime? as it couldnt be running any longer than that, correct?
That is what was there initially (defaultworkertimeout), but as it is defined that any process having a modified value less then `defaultworkertimeout` seconds ago, it is active. That's why I propose to use the existing `findActive()`

Reply:
I meant keeping

    return $this->deleteAll(['modified <' => $thresholdTime]);

and just using the correct max value for thresholdTime

After all, this is just a cleanup function, no? So this shouldnt be a big deal if not done right after the finish of the run.

Reply:
I agree, but my proposal was to use a more dry approach, to say: the cleanup process should remove all not-active processes. So why not use findNotActive, or this case NOT IN findActive. That was the idea, the function of the code does not change.

Reply:
It would only be more DRY if you reuse the query components and not make two queries out of it instead of one.
So best to revert for now, the rest seems mergeable.

Reply:
OK. will revert. But, as I said, the sql server in the php unit test did not like the one query:)

Reply:
Can you clarify what it didnt like?

Reply:
Sure. The query builder was like: `$this->deleteAll(['id NOT IN' => $this->findActive()->select(['id'])]);`

resulting in a query like: `DELETE FROM queue_processes WHERE id NOT IN (SELECT id FROM queue_processes WHERE modified > {THRESHOLDTIME}`



the error was like: Cannot delete from a table while selecting from it at the same time. 
It's not MySQL (as it works on that platform) but probably another service?

Reply:
Where do you see this?
The code seems fine to me: https://github.com/dereuromark/cakephp-queue/blob/5a1f127318210872e3c1a55d210760531fc82dce/src/Model/Table/QueueProcessesTable.php#L204